flink-end-to-end-tests/flink-queryable-state-test/src/main/java/org/apache/flink/streaming/tests/queryablestate/QsStateProducer.java

```diff
@@ -158,24 +161,34 @@ public void open(Configuration parameters) throws Exception {
 
             stateDescriptor.setQueryable(QsConstants.QUERY_NAME);
             state = getRuntimeContext().getMapState(stateDescriptor);
-            updateCount();
-
-            LOG.info("Open {} with a count of {}.", getClass().getSimpleName(), count);
-        }
-
-        private void updateCount() throws Exception {
-            count = Iterables.size(state.keys());
+            countsAtCheckpoint = new HashMap<>();
+            count = -1;
+            lastCompletedCheckpoint = -1;
         }
 
         @Override
         public void flatMap(Email value, Collector<Object> out) throws Exception {
             state.put(value.getEmailId(), new EmailInformation(value));
-            updateCount();
+            count = Iterables.size(state.keys());
         }
 
         @Override
         public void notifyCheckpointComplete(long checkpointId) {
-            System.out.println("Count on snapshot: " + count); // we look for it in the test
+            if (checkpointId > lastCompletedCheckpoint) {
```

Comment:
Is this required to ensure the count (as seen from the test) does not decrease?

Reply:
Yes, it is a safety that that we don't output decreasing values.